velox/exec/NestedLoopJoinProbe.cpp

```diff
@@ -97,60 +352,135 @@ RowVectorPtr NestedLoopJoinProbe::getOutput() {
   }
+  return probeCnt;
+}
 
-  auto size = probeCnt * buildSize;
-  BufferPtr indices = allocateIndices(size, pool());
-  auto* rawIndices = indices->asMutable<vector_size_t>();
+RowVectorPtr NestedLoopJoinProbe::getCrossProduct(
+    vector_size_t probeCnt,
+    const RowTypePtr& outputType,
+    const std::vector<IdentityProjection>& probeProjections,
+    const std::vector<IdentityProjection>& buildProjections) {
+  VELOX_CHECK_GT(probeCnt, 0);
+  VELOX_CHECK(!hasProbedAllBuildData());
+
+  const auto buildSize = buildData_.value()[buildIndex_]->size();
+  const auto batchSize = probeCnt * buildSize;
+  const bool probeCntChanged = (probeCnt != prevProbeCnt_);
+  prevProbeCnt_ = probeCnt;
+  auto output = BaseVector::create<RowVector>(outputType, batchSize, pool());
+
+  auto rawProbeIndices =
+      initializeRowNumberMapping(probeIndices_, batchSize, pool());
   for (auto i = 0; i < probeCnt; ++i) {
     std::fill(
-        rawIndices + i * buildSize,
-        rawIndices + (i + 1) * buildSize,
+        rawProbeIndices.begin() + i * buildSize,
+        rawProbeIndices.begin() + (i + 1) * buildSize,
         probeRow_ + i);
   }
-  auto output = fillOutput(size, indices);
 
-  BufferPtr buildIndices = nullptr;
-  if (probeCnt > 1) {
-    buildIndices = allocateIndices(size, pool());
-    auto* rawBuildIndices = buildIndices->asMutable<vector_size_t>();
+  if (probeCntChanged) {
+    auto rawBuildIndices_ =
+        initializeRowNumberMapping(buildIndices_, batchSize, pool());
     for (auto i = 0; i < probeCnt; ++i) {
       std::iota(
-          rawBuildIndices + i * buildSize,
-          rawBuildIndices + (i + 1) * buildSize,
+          rawBuildIndices_.begin() + i * buildSize,
+          rawBuildIndices_.begin() + (i + 1) * buildSize,
           0);
     }
   }
 
-  auto buildRowVector =
-      buildData_.value()[buildIndex_]->asUnchecked<RowVector>();
-  for (const auto& projection : buildProjections_) {
-    VectorPtr buildVector = buildRowVector->childAt(projection.inputChannel);
-
-    if (buildIndices) {
-      buildVector = BaseVector::wrapInDictionary(
-          BufferPtr(nullptr), buildIndices, size, buildVector);
-    }
-    output->childAt(projection.outputChannel) = buildVector;
-  }
+  projectChildren(output, input_, probeProjections, batchSize, probeIndices_);
+  projectChildren(
+      output,
+      buildData_.value()[buildIndex_],
+      buildProjections,
+      batchSize,
+      buildIndices_);
+  return output;
+}
 
+bool NestedLoopJoinProbe::advanceIndex(vector_size_t probeCnt) {
   probeRow_ += probeCnt;
-  if (probeRow_ == inputSize) {
-    probeRow_ = 0;
+  if (probeRow_ < input_->size()) {
+    return false;
+  }
+  probeRow_ = 0;
+  prevProbeCnt_ = 0;
+  do {
     ++buildIndex_;
-    if (buildIndex_ == buildData_->size()) {
-      buildIndex_ = 0;
-      input_.reset();
+  } while (!hasProbedAllBuildData() &&
+           !buildData_.value()[buildIndex_]->size());
```

Comment:
```
&& buildData_.value()[buildIndex_]->empty()
```

Reply:
`BaseVector` does not have a `empty()` method, should we add one?